common/custom_claims.c

```diff
@@ -0,0 +1,235 @@
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_custom_claims_header_t* header =
+        (oe_custom_claims_header_t*)claims_buffer;
+    size_t claims_index = 0;
+
+    claims_buffer += sizeof(*header);
+    claims_buffer_size -= sizeof(*header);
+    for (uint64_t i = 0; i < header->num_claims; i++)
+    {
+        oe_custom_claims_entry_t* entry =
+            (oe_custom_claims_entry_t*)claims_buffer;
+        uint64_t size = 0;
+
+        // Sanity check sizes.
+        if (claims_buffer_size < sizeof(*entry))
+            OE_RAISE(OE_CONSTRAINT_FAILED);
+
+        OE_CHECK(oe_safe_add_u64(sizeof(*entry), entry->name_size, &size));
+        OE_CHECK(oe_safe_add_u64(size, entry->value_size, &size));
+
+        if (claims_buffer_size < size)
+            OE_RAISE(OE_CONSTRAINT_FAILED);
+
+        // Add the claim.
+        OE_CHECK(_add_claim(
+            &claims[claims_index++],
+            entry->name,
+            entry->name_size,
+            entry->name + entry->name_size,
+            entry->value_size));
+
+        // Go to next entry.
+        claims_buffer += size;
+        claims_buffer_size -= size;
+    }
+
+    result = OE_OK;
+
+done:
+    if (result != OE_OK)
+    {
+        for (size_t i = 0; i < claims_index; i++)
+            _free_claim(&claims[i]);
+    }
+    return result;
+}
+
+oe_result_t oe_deserialize_custom_claims(
+    const uint8_t* claims_buffer,
+    size_t claims_buffer_size,
+    oe_claim_t** claims_out,
+    size_t* claims_length_out)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_custom_claims_header_t* claims_header = NULL;
+    oe_claim_t* claims = NULL;
+    uint64_t claims_length = 0;
+    uint64_t claims_size = 0;
+
+    if (claims_buffer_size < sizeof(oe_custom_claims_header_t))
+    {
+        OE_RAISE(OE_CONSTRAINT_FAILED);
+    }
+
+    claims_header = (oe_custom_claims_header_t*)(claims_buffer);
+
+    if (claims_header == NULL)
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    if (claims_header->version != OE_CUSTOM_CLAIMS_VERSION)
+    {
+        OE_RAISE(OE_CONSTRAINT_FAILED);
+    }
+
+    claims_length = claims_header->num_claims;
+
+    OE_CHECK(oe_safe_mul_u64(claims_length, sizeof(oe_claim_t), &claims_size));
+
+    claims = (oe_claim_t*)oe_malloc(claims_size);
+    if (claims == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    OE_CHECK(_fill_array_with_custom_claims(
+        claims_buffer, claims_buffer_size, claims));
+
+    *claims_out = claims;
+    *claims_length_out = claims_header->num_claims;
+    claims = NULL;
+    result = OE_OK;
+
+done:
+    if (claims)
+        _free_claims(claims, claims_length);
+    return result;
+}
+
+static size_t _get_claims_size(
+    const oe_claim_t* custom_claims,
+    size_t custom_claims_length)
+{
+    size_t size = sizeof(oe_custom_claims_header_t);
+
+    if (!custom_claims)
+        return size;
+
+    for (size_t i = 0; i < custom_claims_length; i++)
+    {
+        size += sizeof(oe_custom_claims_entry_t);
```

Comment:
This is another candidate for safe math. Something you may find helpful is to track in comments any preconditions/postconditions of buffers/sizes being manipulated, particularly for internal functions like this if you're splitting the validation with the caller.